tensorflow/python/eager/function.cc

```diff
@@ -74,10 +84,90 @@ py::object PyConcreteFunction::_build_call_outputs(
   return nest->attr("pack_sequence_as")(structured_outputs, outputs_list, true);
 }
 
+py::object AsNdarray(py::object value) {
+  // TODO(tomhennigan) Support __array_interface__ too.
+  return value.attr("__array__")();
+}
+
+bool IsNdarray(py::object value) {
+  // TODO(tomhennigan) Support __array_interface__ too.
+  PyObject* value_ptr = value.ptr();
+  return PyObject_HasAttr(value_ptr,
+                          PyUnicode_DecodeUTF8("__array__", 9, "strict")) && !(
+      PyObject_IsInstance(value_ptr, Tensor.ptr())
+      || PyObject_IsInstance(value_ptr, BaseResourceVariable.ptr())
+      || PyObject_HasAttr(value_ptr, PyUnicode_DecodeUTF8(
+             "_should_act_as_resource_variable", 32, "strict"))
+      // For legacy reasons we do not automatically promote Numpy strings.
+      || PyObject_IsInstance(value_ptr, np->attr("str_").ptr())
+      // NumPy dtypes have __array__ as unbound methods.
+      || PyObject_IsInstance(value_ptr, (PyObject*) &PyType_Type)
+      // CompositeTensors should be flattened instead.
+      || PyObject_IsInstance(value_ptr, CompositeTensor.ptr()));
+}
+
+py::tuple ConvertNumpyInputs(py::object inputs) {
+  // We assume that any CompositeTensors have already converted their components
+  // from numpy arrays to Tensors, so we don't need to expand composites here for
+  // the numpy array conversion. Instead, we do so because the flattened inputs
+  // are eventually passed to ConcreteFunction()._call_flat, which requires
+  // expanded composites.
+  // TODO(jlchu): Use C++ version of Flatten directly
+  py::list flat_inputs = nest->attr("flatten")(inputs, true);
+
+  // Check for NumPy arrays in arguments and convert them to Tensors.
+  // TODO(nareshmodi): Skip ndarray conversion to tensor altogether, perhaps
+  // finding a way to store them directly in the cache key (currently not
+  // possible since ndarrays are not hashable).
+  bool need_packing = false;
+  py::list filtered_flat_inputs;
+
+  for (int i = 0; i < flat_inputs.size(); ++i) {
+    PyObject* value_ptr = flat_inputs[i].ptr();
+    if (PyObject_IsInstance(value_ptr, Tensor.ptr()) ||
+        PyObject_IsInstance(value_ptr, BaseResourceVariable.ptr())) {
+      filtered_flat_inputs.append(flat_inputs[i]);
+    } else if (PyObject_HasAttr(
```

Comment:
This is a very long condition, can you factor out as a function?

Reply:
I haven't changed it, but the syntax is slightly more readable now. I didn't factor it out originally because it was around half of IsNdarray, and using it there would have invoked another function call. I'm also not sure how to group these conditions under a name. Otherwise, I've left a comment indicating that it is equivalent to the IsNdarray == True case.
Thoughts? Can change if it still seems too dense.